src/search/hnsw_indexer.cc

```diff
@@ -0,0 +1,514 @@
+  }
+  return Status::OK();
+}
+
+Status HnswIndex::RemoveEdge(const NodeKey& node_key1, const NodeKey& node_key2, uint16_t layer,
+                             ObserverOrUniquePtr<rocksdb::WriteBatchBase>& batch) const {
+  auto edge_index_key1 = search_key.ConstructHnswEdge(layer, node_key1, node_key2);
+  auto s = batch->Delete(storage->GetCFHandle(ColumnFamilyID::Search), edge_index_key1);
+  if (!s.ok()) {
+    return {Status::NotOK, fmt::format("failed to delete edge, {}", s.ToString())};
+  }
+
+  auto edge_index_key2 = search_key.ConstructHnswEdge(layer, node_key2, node_key1);
+  s = batch->Delete(storage->GetCFHandle(ColumnFamilyID::Search), edge_index_key2);
+  if (!s.ok()) {
+    return {Status::NotOK, fmt::format("failed to delete edge, {}", s.ToString())};
+  }
+  return Status::OK();
+}
+
+StatusOr<std::vector<VectorItem>> HnswIndex::SelectNeighbors(const VectorItem& vec,
+                                                             const std::vector<VectorItem>& vertors,
+                                                             uint16_t layer) const {
+  std::vector<std::pair<double, VectorItem>> distances;
+  distances.reserve(vertors.size());
+  for (const auto& candidate : vertors) {
+    auto dist = GET_OR_RET(ComputeSimilarity(vec, candidate));
+    distances.emplace_back(dist, candidate);
+  }
+
+  std::sort(distances.begin(), distances.end());
+  std::vector<VectorItem> selected_vs;
+
+  selected_vs.reserve(vertors.size());
+  uint16_t m_max = layer != 0 ? metadata->m : 2 * metadata->m;
+  for (auto i = 0; i < std::min(m_max, (uint16_t)distances.size()); i++) {
+    selected_vs.push_back(distances[i].second);
+  }
+  return selected_vs;
+}
+
+StatusOr<std::vector<VectorItem>> HnswIndex::SearchLayer(uint16_t level, const VectorItem& target_vector,
+                                                         uint32_t ef_runtime,
+                                                         const std::vector<NodeKey>& entry_points) const {
+  std::vector<VectorItem> candidates;
+  std::unordered_set<NodeKey> visited;
+  std::priority_queue<std::pair<double, VectorItem>, std::vector<std::pair<double, VectorItem>>, std::greater<>>
+      explore_heap;
+  std::priority_queue<std::pair<double, VectorItem>> result_heap;
+
+  for (const auto& entry_point_key : entry_points) {
+    Node entry_node = Node(entry_point_key, level);
+    auto entry_node_metadata = GET_OR_RET(entry_node.DecodeMetadata(search_key, storage));
+
+    auto entry_point_vector = VectorItem(entry_point_key, std::move(entry_node_metadata.vector), metadata);
+    auto dist = GET_OR_RET(ComputeSimilarity(target_vector, entry_point_vector));
+
+    explore_heap.push(std::make_pair(dist, entry_point_vector));
+    result_heap.push(std::make_pair(dist, std::move(entry_point_vector)));
+    visited.insert(entry_point_key);
+  }
+
+  while (!explore_heap.empty()) {
+    auto [dist, current_vector] = explore_heap.top();
```

Comment:
```suggestion
    auto [dist, current_vector] = std::move(explore_heap.top());
```

Reply:
Done :)

Reply:
`clang-tidy` throws no-effect error, have to change it back :(
Ref: https://github.com/Beihao-Zhou/kvrocks/actions/runs/9862851656/job/27234342104#step:7:1391

Reply:
Hmm seems here the return type of `top` is const reference so it cannot be moved.